### PR TITLE
add MCS constraints to the SELinux policy

### LINKED ISSUE #1712: enable MCS isolation in the SELinux policy
Currently we allocate random MCS pairs to unprivileged containers, thanks to support in `containerd` and the update to 1.4 in #1336. However, our SELinux policy lacks the constraints required to enforce an isolation boundary.

That means that a process running with the level `s0:c1,c2` can read and write files with the level `s0:c3,c4`, even though those files are nominally owned by a process in a different container.

This will add an additional boundary between containers that supplements the use of namespaces for isolation.

### PR DESCRIPTION
**Pull request by bcressey**
**Issue number:**
Fixes #1712
#1569

**Description of changes:**
This adds `mlsconstrain` statements to the SELinux policy so that the MCS category pairs assigned to container processes and files are used for access decisions.

The CIL documentation on [mlsconstrain](https://github.com/SELinuxProject/selinux/blob/master/secilc/docs/cil_constraint_statements.md#mlsconstrain) is helpful to understand what's happening, but briefly:
* `dom` is the dominance test described in the comments
* `h1` is the source "high" context (typically a subject / process)
* `h2` is the target "high" context (either an object / file, or another subject / process)

The "low" context (`l1`, `l2`) should be just `s0` for everything (with no categories), and is not useful for making the access decision.

Some special cases, since it's hard to write useful comments about what's *not* included:
* `(files (load))` has an exception for "subject" files (like `/proc/<pid>/<file>`), so `ps` works for everyone.
* `(files (describe))` does not have a constraint, so `ls` works for everyone.
* `(files (relabel))` and `(processes (transform))` are "relabel" actions; they're not constrained because they're currently reserved for trusted (read: "not container") processes.

**Testing done:**
Tested these scenarios.

| Action | Privileged container | Unprivileged container |
| --- | --- | --- |
| Create files in an `emptyDir` volume | ✅ | ✅ |
| Create files in a `hostPath` volume | ✅ | ✅ |
| Create files in an Amazon EBS volume | ✅ | ✅ |
| Create files in an Amazon EFS volume | ✅ | ✅ |
| Create files in a Docker image volume | ✅ | ✅ |
| Signal processes with a different level | ✅ | ❌ |
| List files in a directory with a different level | ✅ | ❌ |
| Create files in a directory with a different level | ✅ | ❌ |
| Relabel files with the same level | ✅ | ✅ |
| Relabel files with a different level | ❌ | ❌ |
| Relabel processes with the same level | ✅ | ✅ |
| Relabel processes with a different level | ❌ | ❌ |
| List files in host's `/etc` | ✅ | ❌ |
| Create files in host's `/etc` | ❌ | ❌ |
| List files for host's datastore | ✅ | ❌ |
| Create files in host's datastore | ❌  | ❌ |

Verified that `podman` and `pip` can run inside a container without unexpected denials.

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

### CODE REVIEW

**Review by tjkirch — APPROVED** [review]

**Comment by bcressey** [comment]
Added a new rule to prevent an untrusted process from altering its own range.

**Review by tjkirch — APPROVED** [review]

**Comment by bcressey** [comment]
Modified the rules so that untrusted processes cannot adjust their own labels, even if the resulting value is the same. This aligns with the rules for relabeling files.

**Review by tjkirch — APPROVED** [review]

**Comment by bcressey** [comment]
Fixed a typo spotted by @arnaldo2792.

**Review by arnaldo2792 — APPROVED** [review]
:tada:

**Review by samuelkarp — APPROVED** [review]

**Comment by bcressey** [comment]
Added support for "no-op" relabeling actions where the same label is reapplied. Updated local tests to verify that relabeling actions that do not change the label are allowed, while those that change the label are denied.

**Review by tjkirch — APPROVED** [review]

**Comment by peimanja** [comment]
any update on this one?

**Comment by bcressey** [comment]
It should be merged soon, just waiting on one more review since I reworked it a bit since the reviewers last saw it.

Let me know if there's a test case that you're interested in that you'd like me to try?

**Comment by peimanja** [comment]
> Let me know if there's a test case that you're interested in that you'd like me to try?

Thanks @bcressey! Basically we run github actions on self-hosted runners on Bottlerocket with dind runners for building images.

Workaround you've mentioned in https://github.com/bottlerocket-os/bottlerocket/issues/1569#issuecomment-840101836 works fine but would be nice to get a cleaner way of doing that.

**Review by samuelkarp — APPROVED** [review]
LGTM

**Comment by bcressey** [comment]
> Workaround you've mentioned in [#1569 (comment)](https://github.com/bottlerocket-os/bottlerocket/issues/1569#issuecomment-840101836) works fine but would be nice to get a cleaner way of doing that.

I've been using a version of the pod spec from that issue, with `privileged: true` and without `seLinuxOptions`, and it works fine after these changes.

**Pull request merged**
